### ISSUE TITLE
Undefined behaviour

### ISSUE THREAD

**Issue by efidler** [miloyip/rapidjson#644]
rapidjson relies on undefined behaviour in C++ in several places. I've been trying to make my code that relies on rapidjson run cleanly through ubsan (http://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html) on Clang/OSX and GCC5/Linux.

I'll put up some fixes as pull requests, but I've got some questions about your preferences.

- Do you want each small fix split out?
- Are you ok with using std::numeric_limits?
- Are you ok with compiler annotations to mark deliberate uses of undefined behaviour?
- Are you ok with using compiler intrinsics for checked arithmetic (at least on Clang and GCC)?

**Comment by miloyip** [comment]
>- Do you want each small fix split out?
I am not sure. But separated fixes should be safer and easier for discussion.

>- Are you ok with using std::numeric_limits?
I try to reduce number of header files include. As `cstdint` cannot be prevented, I try to use those macro limits. But if there is template, maybe really need `limits`.

>- Are you ok with compiler annotations to mark deliberate uses of undefined behaviour?
Ideally try to use macro instead of direct use of annotation.

>- Are you ok with using compiler intrinsics for checked arithmetic (at least on Clang and GCC)?
Not sure about this. Maybe similar to the above one.

**Comment by efidler** [comment]
Ok, the pull request is #646.

I split out the individual fixes. I guess one pull request is ok?

I had to use limits, since cstdint wasn't giving me UINT64_MAX or INT64_* on non-C++11 builds.

I put an #if 0 around a couple of tests in valuetest.cpp because they're undefined by nature and I'm not sure what you're trying to test. I fixed the other "deliberate" undefined behaviour.

I ended up removing the __builtin_mul_overflow use, so I didn't need the intrinsic.
